NuRadioMC/utilities/medium_base.py

```diff
@@ -0,0 +1,336 @@
+
+        Returns
+        -------
+        n_nabla:    (3,) np.array
+                    gradient of index of refraction at the point
+        """
+        pass
+
+    
+    def get_ice_model_radiopropa(self):
+        """
+        if radiopropa is installed this will return the ice model to insert in radiopropa
+
+        """
+        if radiopropa_is_imported:
+            # when implementing a new ice_model this part of the function should be ice model specific
+            pass
+        else:
+            logger.error('The radiopropa dependancy was not import and can therefore not be used. \nMore info on https://github.com/nu-radio/RadioPropa')
+            raise ImportError('RadioPropa could not be imported')
+
+
+class IceModel_Simple(IceModel):
+    """
+    predefined ice model (to inherit from) with exponential shape
+    """
+    def __init__(self,n_ice,delta_n,z_0,z_shift=0*units.meter,z_airBoundary=0*units.meter,z_bottom=None):
+        super().__init__(z_airBoundary,z_bottom)
+        self.n_ice = n_ice
+        self.delta_n = delta_n
+        self.z_0 = z_0
+        self.z_shift = z_shift
+
+    def get_index_of_refraction(self, x):
+        """
+        overwrite inherited function
+        """
+        if (x[2] - self.z_airBoundary) <=0:
+            return self.n_ice  - self.delta_n  * np.exp((x[2]-self.z_shift) / self.z_0)
+        else:
+            return 1
+
+    def get_average_index_of_refraction(self, x1, x2):
+        """
+        overwrite inherited function
+        """
+        if ((x1[2] - self.z_airBoundary) <=0) and ((x2[2] - self.z_airBoundary) <=0):
+            return self.n_ice - self.delta_n * self.z_0 / (x2[2] - x1[2]) * (np.exp((x2[2]-self.z_shift) / self.z_0) - np.exp((x1[2]-self.z_shift) / self.z_0))
+        else:
+            return None
+
+    def get_gradient_of_index_of_refraction(self, x):
+        """
+        overwrite inherited function
+        """
+        gradient = np.array([0.,0.,0.])
+        if (x[2] - self.z_airBoundary) <=0:
+            gradient[2] = - self.delta_n / self.z_0 * np.exp((x[2]-self.z_shift) / self.z_0)
+        return gradient
+
+    def get_ice_model_radiopropa(self):
+        """
+        overwrite inherited function
+        """
+        if radiopropa_is_imported:
+            scalar_field = RP.IceModel_Simple(z_surface=self.z_airBoundary*RP.meter/units.meter, 
+                                            n_ice=self.n_ice, delta_n=self.delta_n, 
+                                            z_0=self.z_0*RP.meter/units.meter,
+                                            z_shift=self.z_shift*RP.meter/units.meter)
+            return RadioPropaIceWrapper(self,scalar_field)
+        else:
+            logger.error('The radiopropa dependancy was not import and can therefore not be used. \nMore info on https://github.com/nu-radio/RadioPropa')
+            raise ImportError('RadioPropa could not be imported')
+
+
+
+if radiopropa_is_imported:
+    class RadioPropaIceWrapper():
```

Comment:
You mentioned that ice models need to be explicitly reimplemented in RadioPropa to avoid a (Python) performance penalty. Does this apply for simple ice models also, where the only relevant information is a couple of floats? (I guess this is not super relevant, as one should use the analytic ray tracer for these ice models anyway).

Having to implement ice models separately in NuRadioMC (e.g. for the parabolic equation solver, or other future propagation modules that can also work with scalar fields) and in radiopropa also seems a bit unfortunate - I would have imagined maybe an explicit scalar field array that could be passed on to the propagation module, so one doesn't have the constant python function calls slowing down a propagation module that's written in c++. I guess the downside would be a (large?) increase in memory usage, plus maybe a bunch of other issues that you have already thought of.

Reply:
For simple ice models this is not the case because there already is a RP object that holds all the functions for simple ice model, it just needs to have the right floats and then the NuRadioMC class handles everything.

Reply:
Re: simple ice models: excellent! That's what I hoped was the case

Wrt scalar field ice models: right now, my understanding is that if you have a scalar field that's implemented in NuRadioMC, but not in RadioPropa, it will 'wrap' the scalar field using ScalarFieldBuilder. This means that when you run the radiopropa solver, at every position it evaluates the index of refraction, a separate call is issued to Python, which slows things down massively. It would be quicker if you could just generate a big grid (array) of potential positions, ask the Python ice model **once** for the index of refraction at all these grid points, and then pass a big array of floats to RadioPropa, after which no more calls to Python would be necessary.
As mentioned though, this would take up a lot more memory (potentially way too much, if the grid spacing is very small - but I feel like we should be limited by our resolution on the ice density anyway?), and also depends on how this is implemented inside RadioPropa.

Reply:
Ah, okay, I understand now!
It is a clever suggestion indeed. For the moment that is not straight forward to do I think because of the implementation of radiopropa, but I definitely take it up for a next iteration on the module :)